_posts/2019-12-18-package-lock.md

```diff
@@ -0,0 +1,101 @@
+
+<pre><code>npm notice created a lockfile as package-lock.json. You should commit this file.
+</code></pre>
+
+While this advice is well intentioned...  it's not always true :) Let's talk about when you want to check this into source control, and when not to.
+
+## How package-lock.json works
+
+The mechanics of `package-lock.json` are simple enough:
+- When you run `npm install` the first time, npm automatically creates the file for you.
+- It has the exact version of every top level and transitive dependency in your `node_modules` folder.
+- When you update the version of a package in your `package.json` and run `npm install`, the `package-lock.json` file will get updated automatically.
+
+If you peek inside one of these files, you're going to get a bunch of non-human-readable gobbledygook like this:
+
+<pre><code class="language-js">
+{
+  "name": "@google-cloud/bigtable",
+  "version": "2.3.0",
+  "lockfileVersion": 1,
+  "requires": true,
+  "dependencies": {
+    "@babel/code-frame": {
+      "version": "7.5.5",
+      "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.5.5.tgz",
+      "integrity": "sha512-27d4lZoomVyo51VegxI20xZPuSHusqbQag/ztrBC7wegWoQ1nLREPVSKSW8byhTlzTKyNE4ifaTA6lCp7JjpFw==",
+      "dev": true,
+      "requires": {
+        "@babel/highlight": "^7.0.0"
+      }
+    },
+  ...
+</code></pre>
+
+The next time you run `npm install`, you're (likely) going to get the exact copy of dependencies outlined in the `package-lock.json` file, instead of going through the module resolution dance from the contents of `package.json`.
+
+## Why lock files sound great
+In a world where lock files do not exist - running `npm install` is largely [non-idempotent](https://en.wikipedia.org/wiki/Idempotence).  This means that I can run `npm install` twice, and the results in the `node_modules` folder can be wildly different.  Stop me if you've seen this happen before:
+- *I just wrote some code locally, and everything works*
+- *I go ahead and `git push` my changes.  For some reason CI is breaking!?*
+- *It's a lint error... it looks like `prettier` is failing on some formatting change*.
+
+This can happen because **the version of `prettier` you're using locally can be different than the version of `prettier` you're getting in CI**.  This is one of the critical problems package lock files can solve - the version you're getting on your computer, and on CI are (mostly) going to be the same.
+
+Having an exact record of the full transitive dependency tree can be great for a few reasons:
+- Every environment is (mostly) guaranteed to install the same set of dependencies when running `npm install`
+- `npm install` typically runs faster because some resolution steps can be skipped
+- There is a log in source control of the exact tree used to run your code and tests
+
+This all sounds great, but guess what ... *I don't use them, and if you're building libraries, you probably shouldn't use them either*.
```

Comment:
s/them/lock files/

(you have a lot of nouns above that "them" could refer to)

Reply:
👍